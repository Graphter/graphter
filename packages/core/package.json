{
  "name": "@graphter/core",
  "version": "1.0.0",
  "description": "",
  "main": "dist/index.js",
  "module": "dist/index.es.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "scripts": {
    "test": "jest --config jest.config.js",
    "test:watch": "react-scripts test --env=jsdom",
    "test:update-snapshots": "jest --updateSnapshot",
    "build": "rollup -c",
    "start": "rollup -c -w",
    "prepare": "npm run build"
  },
  "author": "Jonathan Pritchard",
  "license": "ISC",
  "peerDependencies": {
    "react": "^16.13.1",
    "recoil": "0.0.8"
  },
  "dependencies": {},
  "devDependencies": {
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-export-default-from": "^7.8.3",
    "@babel/preset-env": "^7.9.0",
    "@babel/preset-stage-0": "^7.8.3",
    "@rollup/plugin-json": "^4.0.3",
    "@types/jest": "^25.2.1",
    "@types/react": "^16.9.49",
    "@types/recoil": "0.0.1",
    "ajv": "^6.12.2",
    "eslint": "^6.8.0",
    "jest": "^25.5.4",
    "rollup": "^2.2.0",
    "rollup-plugin-babel": "^4.4.0",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-json": "^4.0.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.2",
    "rollup-plugin-typescript2": "^0.26.0",
    "ts-jest": "^25.3.1",
    "typescript": "^3.8.3"
  }
}
