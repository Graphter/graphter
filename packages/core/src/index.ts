export * from './ListDisplayConfig'
export * from './Service'
export * from './ListResult'
export * from './GetResult'
export * from './SaveResult'
export * from './NodeConfig'
export * from './NodeRendererProps'
export * from './ErrorRendererProps'
export * from './NodeRendererRegistration'
export * from './JsonType'
export * from './PathSegment'
export * from './NodeValidation'
export * from './ValidationErrorDisplalyMode'
export * from './ValidationExecutionStage'
export * from './ValidationResult'
export * from './NodeValidatorRegistration'