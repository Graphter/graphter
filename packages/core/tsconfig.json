{
  "compilerOptions": {
    "declaration": true,
    "declarationDir": "./dist",
    "module": "esnext",
    "outDir": "./dist",
    "target": "es5",
    "jsx": "react",
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "strictNullChecks": true,
    "allowSyntheticDefaultImports": true,
    "moduleResolution": "node",
    "allowJs": true,
    "sourceMap": true,
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "skipLibCheck": true,
    "esModuleInterop": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true
  },
  "include": [
    "src/**/*"
  ],
  "exclude": [
    "node_modules"
  ]
}
