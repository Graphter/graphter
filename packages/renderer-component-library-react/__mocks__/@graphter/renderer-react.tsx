import React, { useState } from "react";

export const useArrayNodeData = jest.fn(() => {
  return {
    childIds: [],
    removeItem: jest.fn(),
    commitItem: jest.fn()
  }
})

export const useNodeData = jest.fn(() => {
  return [
    null,
    () => {}
  ]
})

export const useNodeValidation = jest.fn().mockResolvedValue([])

export const nodeRendererStore = {
  get: jest.fn((nodeType: string) => ({
    type: 'string',
    renderer: ({originalNodeData}: any) => {
      const [ value, setValue ] = useState(originalNodeData || '')
      return (
        <input value={value} onChange={e => setValue(e.target.value)} data-testid='string-renderer' />
    )
    }
  }))
}
export const createDefault = jest.fn()