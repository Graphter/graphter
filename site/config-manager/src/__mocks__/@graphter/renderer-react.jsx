"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pathConfigStore = exports.createDefault = exports.nodeRendererStore = exports.useNodeValidation = exports.useNodeData = exports.useArrayNodeData = void 0;
var react_1 = require("react");
exports.useArrayNodeData = jest.fn(function () {
    return {
        childIds: [],
        removeItem: jest.fn(),
        commitItem: jest.fn()
    };
});
exports.useNodeData = jest.fn(function () {
    return [
        null,
        function () { }
    ];
});
exports.useNodeValidation = jest.fn().mockResolvedValue([]);
exports.nodeRendererStore = {
    get: jest.fn(function (nodeType) { return ({
        type: 'string',
        renderer: function (_a) {
            var originalNodeData = _a.originalNodeData;
            var _b = react_1.useState(originalNodeData || ''), value = _b[0], setValue = _b[1];
            return (<input value={value} onChange={function (e) { return setValue(e.target.value); }} data-testid='string-renderer'/>);
        }
    }); })
};
exports.createDefault = jest.fn();
exports.pathConfigStore = {
    get: jest.fn(),
    set: jest.fn()
};
