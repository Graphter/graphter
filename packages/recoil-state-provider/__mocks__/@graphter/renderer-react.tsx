import { NodeRendererRegistration } from "@graphter/core";
import React from "react";

export const nodeRendererStore = {
  get: jest.fn()
}

export const pathConfigStore = {
  get: jest.fn(),
  set: jest.fn()
}

export const NodeValidationProvider = jest.fn(({ children }: any) => <div>{children}</div>)
export const NodeDataProvider = jest.fn(({ children }: any) => <div>{children}</div>)