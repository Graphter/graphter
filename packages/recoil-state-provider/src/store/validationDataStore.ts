import { NodeConfig, PathSegment, ValidationResult } from "@graphter/core";
import { atom, RecoilState, RecoilValueReadOnly, selector } from "recoil";
import { NodeValidationData } from "@graphter/renderer-react";
import { propDataStore } from "./propDataStore";

const validationStateMap: Map<string, RecoilState<NodeValidationData>> = new Map()
const pathSegmentKeySalt = '[be3cc399-9225-4cbc-9d23-6c0c69ddca4a]'
const aggregateValidationsStateMap: Map<string, RecoilValueReadOnly<Array<NodeValidationData>>> = new Map()
const pathKeySalt = '[c42e99e6-d234-4d4a-828f-f696b11c9683]'

const generateValidationKey = (path: Array<PathSegment>) => `validation-${path.join(pathSegmentKeySalt)}`

export const get = (path: Array<PathSegment>): RecoilState<NodeValidationData> => {
  const state = validationStateMap.get(generateValidationKey(path))
  if(!state) throw new Error(`No validation state for '${path.join('/')}' found in the store`)
  return state
}

export const getAll = (paths: Array<Array<PathSegment>>): RecoilValueReadOnly<Array<NodeValidationData>> => {
  const key = `aggregate-validations-${paths.map(path => path.join(pathSegmentKeySalt)).join(pathKeySalt)}`
  if(aggregateValidationsStateMap.has(key)){
    const aggregateValidations = aggregateValidationsStateMap.get(key)
    if(!aggregateValidations) throw new Error('Could not find aggregate validation. Should never happen')
    return aggregateValidations
  }

  const aggregateValidations = selector<Array<NodeValidationData>>({
    key,
    get: ({ get: selectorGet }) => {
      return paths.flatMap(path => {
        if(!has(path)) {
          if(!propDataStore.has(path)) return []
          const propState = propDataStore.get(path)
          const propData = selectorGet(propState)
          set(path, propData, [])
        }
        const validationDataState = get(path)
        const validationData = selectorGet(validationDataState)
        return [{
          path,
          results: validationData.results
        }]
      })
    }
  })
  aggregateValidationsStateMap.set(key, aggregateValidations)

  return aggregateValidations
}

export const has = (path: Array<PathSegment>): boolean => {
  return validationStateMap.has(generateValidationKey(path))
}

export const set = (
  path: Array<PathSegment>,
  value: any,
  results: Array<ValidationResult>
) => {
  if(!path || !path.length) throw new Error('path argument is required')
  const key = generateValidationKey(path)
  const validationData: NodeValidationData = {
    path,
    value,
    results
  }
  validationStateMap.set(key, atom({
    key: key,
    default: validationData
  }))
}

export interface ValidationDataStore{
  get: (path: Array<PathSegment>) => RecoilState<NodeValidationData>
  getAll: (paths: Array<Array<PathSegment>>) => RecoilValueReadOnly<Array<NodeValidationData>>
  has: (path: Array<PathSegment>) => boolean
  set: (
    path: Array<PathSegment>,
    value: any,
    results: Array<ValidationResult>
  ) => void
}

const validationDataStore: ValidationDataStore = {
  get,
  getAll,
  has,
  set
}

export default validationDataStore