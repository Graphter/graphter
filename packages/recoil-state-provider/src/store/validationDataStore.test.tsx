import { ValidationDataStore } from "./validationDataStore";
import { RecoilRoot, useRecoilValue } from "recoil";
import { render } from "@testing-library/react";
import React from "react";
import { PathSegment } from "@graphter/core";
import PropDataStore from "./propDataStore";

describe('validationDataStore', () => {
  let validationDataStore: ValidationDataStore,
    propDataStore: PropDataStore

  function GetConsumerComponent({ cb }: { cb?: (data: any) => void }){
    const validationData = useRecoilValue(validationDataStore.get(['page']))
    if(cb) cb(validationData)
    return null
  }

  beforeEach(() => {
    jest.isolateModules(() => {
      validationDataStore = require('./validationDataStore')
      propDataStore = require('./propDataStore')
    })
  })
  it('should set and get validation data', () => {
    validationDataStore.set(
      ['page'],
      'the-page-value',
      [ { valid: false, errorMessage: 'Some error' }]
      )
    const cbMock = jest.fn()
    render(
      <RecoilRoot>
        <GetConsumerComponent cb={cbMock} />
      </RecoilRoot>
    )
    expect(cbMock).toHaveBeenCalled()
    expect(cbMock.mock.calls[0][0]).toMatchSnapshot()
  })
  describe('get()', () => {
    it('should error if no validation data has been set yet', () => {
      expect(() => render(
        <RecoilRoot>
          <GetConsumerComponent />
        </RecoilRoot>
      )).toThrowErrorMatchingSnapshot()
    })
  })
  describe('getAll()', () => {
    function GetAllConsumerComponent({ paths, cb }: {
      paths: Array<Array<PathSegment>>
      cb?: (data: any) => void
    }){
    const validationData = useRecoilValue(validationDataStore.getAll(paths))
    if(cb) cb(validationData)
    return null
  }
    it('should return a selector that aggregates results for all the supplied paths', () => {
      validationDataStore.set(
        ['page'],
        'the-page-value',
        [ { valid: false, errorMessage: 'Some page error' }]
      )
      validationDataStore.set(
        ['page', 'title'],
        'the-title-value',
        [ { valid: false, errorMessage: 'Some title error' }]
      )
      const cbMock = jest.fn()
      render(
        <RecoilRoot>
          <GetAllConsumerComponent paths={[ ['page'], ['page', 'title'] ]} cb={cbMock} />
        </RecoilRoot>
      )
      expect(cbMock).toHaveBeenCalled()
      expect(cbMock.mock.calls[0][0]).toMatchSnapshot()
    })
    it('should cache the selector for subsequent uses', () => {
      validationDataStore.set(
      ['page'],
      'the-page-value',
      [ { valid: false, errorMessage: 'Some page error' }]
      )
      validationDataStore.set(
        ['page', 'title'],
        'the-title-value',
        [ { valid: false, errorMessage: 'Some title error' }]
      )
      const selector1 = validationDataStore.getAll([ ['page'], ['page', 'title'] ])
      const selector2 = validationDataStore.getAll([ ['page'], ['page', 'title'] ])
      expect(selector1).toBe(selector2)
    })
    it('should initialise state on the fly when not already initialised', () => {
      propDataStore.set(
        ['page'],
        true,
        'the-page-value'
      )
      propDataStore.set(
        ['page', 'title'],
        true,
        'the-title-value',
      )
      const cbMock = jest.fn()
      render(
        <RecoilRoot>
          <GetAllConsumerComponent paths={[ ['page'], ['page', 'title'] ]} cb={cbMock} />
        </RecoilRoot>
      )
      expect(cbMock).toHaveBeenCalled()
      expect(cbMock.mock.calls[0][0]).toMatchSnapshot()
    })
  })
  describe('has()', () => {
    it('should return true if validation data set been set', () => {
      validationDataStore.set(
        ['page'],
        'the-page-value',
        [ { valid: false, errorMessage: 'Some page error' }]
      )
      expect(validationDataStore.has(['page'])).toBe(true)
    })
    it('should return false if validation has not been set', () => {
      expect(validationDataStore.has(['page'])).toBe(false)
    })
  })
  describe('set()', () => {
    it('should error if no path is supplied', () => {
      expect(() => validationDataStore.set(
        // @ts-ignore
        null,
        'the-page-value',
        []
      )).toThrowErrorMatchingSnapshot()
    })
  })
})
export {}