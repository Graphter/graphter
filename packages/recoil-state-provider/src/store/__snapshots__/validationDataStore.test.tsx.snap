// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validationDataStore get() should error if no validation data has been set yet 1`] = `"No validation state for 'page' found in the store"`;

exports[`validationDataStore getAll() should initialise state on the fly when not already initialised 1`] = `
Array [
  Object {
    "path": Array [
      "page",
    ],
    "results": Array [],
  },
  Object {
    "path": Array [
      "page",
      "title",
    ],
    "results": Array [],
  },
]
`;

exports[`validationDataStore getAll() should return a selector that aggregates results for all the supplied paths 1`] = `
Array [
  Object {
    "path": Array [
      "page",
      "title",
    ],
    "results": Array [
      Object {
        "errorMessage": "Some title error",
        "valid": false,
      },
    ],
  },
  Object {
    "path": Array [
      "page",
      "author",
      "location",
    ],
    "results": Array [
      Object {
        "errorMessage": "Some author location error",
        "valid": false,
      },
    ],
  },
]
`;

exports[`validationDataStore set() should error if no path is supplied 1`] = `"path argument is required"`;

exports[`validationDataStore should set and get validation data 1`] = `
Object {
  "path": Array [
    "page",
  ],
  "results": Array [
    Object {
      "errorMessage": "Some error",
      "valid": false,
    },
  ],
  "value": "the-page-value",
}
`;
